src/main/java/com/google/devtools/build/remote/client/ActionGrouping.java

```diff
@@ -43,17 +104,98 @@ public void addLogEntry(LogEntry entry) {
                 // In the improbable case of the same timestamp, ensure the messages do not
                 // override each other.
                 return a.hashCode() - b.hashCode();
-              }));
+              });
+      summary = new ActionResultSummary(actionId);
+    }
+
+    private Digest extractDigest(LogEntry entry) {
+      if (!entry.hasDetails()) {
+        return null;
+      }
+      RpcCallDetails details = entry.getDetails();
+      if (details.hasExecute()) {
+        if (details.getExecute().hasRequest()
+            && details.getExecute().getRequest().hasActionDigest()) {
+          return details.getExecute().getRequest().getActionDigest();
+        }
+      }
+      if (details.hasGetActionResult()) {
+        if (details.getGetActionResult().hasRequest()
+            && details.getGetActionResult().getRequest().hasActionDigest()) {
+          return details.getGetActionResult().getRequest().getActionDigest();
+        }
+      }
+      return null;
+    }
+
+    Digest getDigest() {
+      return digest;
+    }
+
+    void add(LogEntry entry) throws IOException {
+      log.add(entry);
+
+      Digest d = extractDigest(entry);
+      if (d != null) {
+        if (digest != null && !d.equals(digest)) {
+          System.err.println("Warning: conflicting digests: " + d + " and " + digest);
+        }
+        digest = d;
+      }
+
+      summary.add(entry);
+    }
+
+    // We will consider an action to be failed if we successfully got an action result but the exit
+    // code is non-zero
+    boolean isFailed() {
+      return summary.getActionResult() != null && summary.getActionResult().getExitCode() != 0;
+    }
+
+    Iterable<? extends LogEntry> getSortedElements() {
+      return log;
```

Comment:
Consider using the Builder pattern for this whole class. In general, it is a custom in Java (and in Bazel in particular) that classes are immutable after creation. This has a number of advantages, primarily much stricter invariants (don't need to check all the time that X happens only after Y). Here in particular, if you extract all the mutations/construction into the ActionDetails.Builder class, and have ActionDetails itself be immutable, you can concentrate all correctness checks in one logical place.

Reply:
Good idea!
I've rewritten the classes using the Builder pattern and moved out some of the static functions into the Utils, I agree it reads cleaner now